File: firmware/core/fseq.c
/* Leitor fseq V2 — ver fseq.h. Layout conferido contra o escritor
   do sequenciador (web/src/motor/fseq.js) e o FPP (FSEQFile.cpp);
   os testes de host leem arquivos gerados pelo exportador real. */

#include "fseq.h"

#include <stdio.h>
#include <string.h>
#include <zlib.h>

static uint16_t u16(const uint8_t *p) { return (uint16_t)(p[0] | p[1] << 8); }
static uint32_t u32(const uint8_t *p) {
  return (uint32_t)p[0] | (uint32_t)p[1] << 8 |
         (uint32_t)p[2] << 16 | (uint32_t)p[3] << 24;
}
static uint64_t u64(const uint8_t *p) {
  return (uint64_t)u32(p) | (uint64_t)u32(p + 4) << 32;
}

#define CABECALHO 32
#define BLOCO_INDICE 8
#define FAIXA_ESPARSA 6
#define VAR_CABECALHO 4

int fseq_ler_cabecalho(const uint8_t *buf, size_t len, fseq_info_t *fi) {
  memset(fi, 0, sizeof(*fi));
  if (len < CABECALHO) return FSEQ_ERRO_CURTO;
  if (memcmp(buf, "PSEQ", 4) != 0) return FSEQ_ERRO_MAGIC;
  if (buf[7] != 2) return FSEQ_ERRO_VERSAO;

  fi->offset_dados = u16(buf + 4);
  fi->canais = u32(buf + 10);
  fi->quadros = u32(buf + 14);
  fi->passo_ms = buf[18];
  fi->compressao = buf[20] & 0x0F;
  /* 12 bits partidos: nibble alto do byte 20 são os bits altos.
     É o detalhe do formato que mais dá erro de implementação. */
  fi->n_blocos = (uint16_t)(((buf[20] & 0xF0) << 4) | buf[21]);
  fi->unique_id = u64(buf + 24);

  if (fi->compressao == FSEQ_ZSTD) return FSEQ_ERRO_COMPRESSAO;
  if (fi->compressao > FSEQ_ZLIB) return FSEQ_ERRO_COMPRESSAO;
  if (fi->canais == 0 || fi->quadros == 0) return FSEQ_ERRO_FORMATO;
  if (len < fi->offset_dados) return FSEQ_ERRO_CURTO;

  size_t p = CABECALHO + (size_t)fi->n_blocos * BLOCO_INDICE
           + (size_t)buf[22] * FAIXA_ESPARSA;
  if (p > fi->offset_dados) return FSEQ_ERRO_FORMATO;

  /* Cabeçalhos variáveis: u16 tamanho (contando estes 4 bytes),
     2 bytes de código, dado. Código desconhecido se ignora — é a
     regra do formato, e é o que nos deixa carimbar o "bl". */
  while (p + VAR_CABECALHO <= fi->offset_dados) {
    uint16_t tam = u16(buf + p);
    if (tam < VAR_CABECALHO || p + tam > fi->offset_dados) break;
    const uint8_t *dado = buf + p + 4;
    size_t n = tam - VAR_CABECALHO;
    while (n > 0 && dado[n - 1] == 0) n--;         /* tira o NUL do fim */

    if (buf[p + 2] == 'm' && buf[p + 3] == 'f') {
      size_t c = n < sizeof(fi->midia) - 1 ? n : sizeof(fi->midia) - 1;
      memcpy(fi->midia, dado, c);
      fi->midia[c] = 0;
    } else if (buf[p + 2] == 'b' && buf[p + 3] == 'l') {
      /* "versão;canais;impressão" — versão que não conhecemos se
         ignora: melhor tocar com aviso do que reprovar arquivo bom. */
      char txt[48];
      size_t c = n < sizeof(txt) - 1 ? n : sizeof(txt) - 1;
      memcpy(txt, dado, c);
      txt[c] = 0;
      unsigned versao = 0, canais = 0;
      char fp[16] = { 0 };
      if (sscanf(txt, "%u;%u;%15[0-9a-fA-F]", &versao, &canais, fp) == 3 &&
          versao == 1) {
        fi->tem_rig = true;
        fi->rig_canais = canais;
        memcpy(fi->rig_fp, fp, sizeof(fp));
      }
    }
    p += tam;
  }
  return FSEQ_OK;
}

int fseq_bloco_do_quadro(const uint8_t *buf, size_t len,
                         const fseq_info_t *fi, uint32_t q,
                         fseq_bloco_t *bl) {
  if (q >= fi->quadros) return FSEQ_ERRO_FORA;

  if (fi->compressao == FSEQ_SEM_COMPRESSAO) {
    bl->primeiro_quadro = 0;
    bl->quadros = fi->quadros;
    bl->offset = fi->offset_dados;
    bl->tam = fi->canais * fi->quadros;
    return FSEQ_OK;
  }

  if (len < (size_t)CABECALHO + (size_t)fi->n_blocos * BLOCO_INDICE)
    return FSEQ_ERRO_CURTO;

  uint32_t off = fi->offset_dados;
  for (uint16_t i = 0; i < fi->n_blocos; i++) {
    const uint8_t *e = buf + CABECALHO + (size_t)i * BLOCO_INDICE;
    uint32_t primeiro = u32(e);
    uint32_t tam = u32(e + 4);
    if (tam == 0) continue;                        /* entrada vazia: pula */
    uint32_t fim = fi->quadros;
    /* fim do bloco = primeiro quadro do próximo bloco não-vazio */
    for (uint16_t j = i + 1; j < fi->n_blocos; j++) {
      const uint8_t *e2 = buf + CABECALHO + (size_t)j * BLOCO_INDICE;
      if (u32(e2 + 4) > 0) { fim = u32(e2); break; }
    }
    if (q >= primeiro && q < fim) {
      bl->primeiro_quadro = primeiro;
      bl->quadros = fim - primeiro;
      bl->offset = off;
      bl->tam = tam;
      return FSEQ_OK;
    }
    off += tam;
  }
  return FSEQ_ERRO_FORA;
}

int fseq_offset_quadro(const fseq_info_t *fi, uint32_t q, uint32_t *off) {
  if (fi->compressao != FSEQ_SEM_COMPRESSAO) return FSEQ_ERRO_FORMATO;
  if (q >= fi->quadros) return FSEQ_ERRO_FORA;
  *off = fi->offset_dados + q * fi->canais;
  return FSEQ_OK;
}

int fseq_descomprime(const fseq_info_t *fi, const fseq_bloco_t *bl,
                     const uint8_t *comprimido, uint8_t *saida,
                     size_t saida_cap) {
  size_t esperado = (size_t)bl->quadros * fi->canais;
  if (saida_cap < esperado) return FSEQ_ERRO_BUFFER;

  if (fi->compressao == FSEQ_SEM_COMPRESSAO) {
    memcpy(saida, comprimido, esperado);
    return FSEQ_OK;
  }

  /* zlib RFC1950 — o mesmo que o CompressionStream("deflate") do
     browser escreve e o deflateInit() do FPP produz. */
  uLongf tam = (uLongf)saida_cap;
  if (uncompress(saida, &tam, comprimido, bl->tam) != Z_OK)
    return FSEQ_ERRO_ZLIB;
  if (tam != esperado) return FSEQ_ERRO_ZLIB;
  return FSEQ_OK;
}

int fseq_conferir_rig(const fseq_info_t *fi, uint32_t canais_atual,
                      const char *fp_atual) {
  if (!fi->tem_rig) return FSEQ_RIG_SEM_CARIMBO;
  if (fi->rig_canais != canais_atual) return FSEQ_RIG_CANAIS;
  if (strncmp(fi->rig_fp, fp_atual, sizeof(fi->rig_fp)) != 0)
    return FSEQ_RIG_CROQUI;
  return FSEQ_RIG_OK;
}

const char *fseq_erro(int c) {
  switch (c) {
    case FSEQ_OK: return "ok";
    case FSEQ_ERRO_CURTO: return "buffer curto";
    case FSEQ_ERRO_MAGIC: return "nao e fseq";
    case FSEQ_ERRO_VERSAO: return "so V2";
    case FSEQ_ERRO_COMPRESSAO: return "compressao nao suportada";
    case FSEQ_ERRO_FORMATO: return "cabecalho inconsistente";
    case FSEQ_ERRO_FORA: return "quadro fora do arquivo";
    case FSEQ_ERRO_ZLIB: return "zlib falhou";
    case FSEQ_ERRO_BUFFER: return "buffer de saida pequeno";
    default: return "erro desconhecido";
  }
}

File: web/src/motor/fseq.js
/* ============================================================
   FSEQ V2 — escrita e leitura.

   Formato aberto do FPP/xLights. Escrever nele é o que tira o
   sequenciador da tela e põe luz no carro, e de quebra deixa usar
   o xLights como validador: sequencia lá, exporta, compara aqui.

   Layout do cabeçalho (32 bytes), conferido contra
   FalconChristmas/fpp — src/fseq/FSEQFile.cpp:

     0..3   "PSEQ"
     4..5   u16  offset onde começa o dado de canal
     6      u8   versão menor
     7      u8   versão maior (2)
     8..9   u16  tamanho do cabeçalho: 32 + blocos*8 + esparsos*6
     10..13 u32  canais por quadro
     14..17 u32  número de quadros
     18     u8   passo em ms
     19     u8   flags (0)
     20     u8   (blocos >> 4 & 0xF0) | tipo de compressão
     21     u8   blocos & 0xFF
     22     u8   número de faixas esparsas
     23     u8   reservado (0)
     24..31 u64  id único
     -->    índice de blocos: blocos × (u32 primeiro quadro, u32 bytes)
     -->    faixas esparsas: esparsos × (u24 canal inicial, u24 tamanho)
     -->    cabeçalhos variáveis: u16 tamanho (com estes 4 bytes), 2 bytes de
            código, dado. "mf" é o nome do arquivo de mídia
     -->    dado de canal, alinhado em múltiplo de 4
   ============================================================ */

export const COMPRESSAO = { nenhuma: 0, zstd: 1, zlib: 2 };
const NOME_COMPRESSAO = ["nenhuma", "zstd", "zlib"];

const CABECALHO = 32;
const BLOCO_INDICE = 8;
const FAIXA_ESPARSA = 6;
const VAR_CABECALHO = 4;

/* Alvo de 64KB descomprimidos por bloco, o mesmo do FPP. Bloco é a
   unidade de leitura do player: menor que isso desperdiça índice,
   maior faz o ESP segurar RAM à toa e aumenta o custo de um seek. */
const BLOCO_ALVO = 64 * 1024;
const MAX_BLOCOS = 255;

/* Carimbo do croqui: `versão;canais;impressão`. Código de duas letras
   nosso, no espaço de cabeçalho variável do próprio formato. */
export const CODIGO_RIG = "bl";
const VERSAO_RIG = 1;

const arredonda4 = v => (v + 3) & ~3;

/* zlib pelo padrão da plataforma: CompressionStream existe no browser
   e no Node 18+. "deflate" aqui é RFC1950 (com cabeçalho zlib), que é
   exatamente o que o deflateInit() do FPP produz e espera. */
async function zlibComprime(u8) {
  const s = new Blob([u8]).stream().pipeThrough(new CompressionStream("deflate"));
  return new Uint8Array(await new Response(s).arrayBuffer());
}

async function zlibDescomprime(u8) {
  const s = new Blob([u8]).stream().pipeThrough(new DecompressionStream("deflate"));
  return new Uint8Array(await new Response(s).arrayBuffer());
}

/* Quantos quadros cabem num bloco, respeitando o teto de blocos que
   o cabeçalho consegue endereçar. */
export function quadrosPorBloco(canais, quadros) {
  let porBloco = Math.max(2, Math.floor(BLOCO_ALVO / Math.max(1, canais)));
  porBloco = Math.min(porBloco, Math.max(2, quadros));
  while (Math.ceil(quadros / porBloco) > MAX_BLOCOS) porBloco++;
  return porBloco;
}

/**
 * Monta o arquivo .fseq inteiro na memória.
 *
 * @param {object} o
 * @param {number} o.canais       canais por quadro
 * @param {number} o.quadros      número de quadros
 * @param {Uint8Array} o.dados    quadros × canais bytes, quadro a quadro
 * @param {number} [o.stepTimeMs] passo entre quadros (25 = 40fps)
 * @param {"nenhuma"|"zlib"} [o.compressao]
 * @param {bigint} [o.uniqueId]   id do arquivo; padrão é o relógio
 * @param {string} [o.midia]      nome do arquivo de áudio ("mf")
 * @param {{ch:number,fp:string}} [o.rig] carimbo do croqui ("bl")
 * @returns {Promise<Uint8Array>}
 */
export async function escreverFseq({
  canais, quadros, dados, stepTimeMs = 25,
  compressao = "zlib", uniqueId, midia, rig,
}) {
  if (!Number.isInteger(canais) || canais <= 0) throw new Error("canais inválido");
  if (!Number.isInteger(quadros) || quadros <= 0) throw new Error("quadros inválido");
  if (dados.length !== canais * quadros)
    throw new Error(`dados tem ${dados.length} bytes, esperado ${canais * quadros}`);
  if (stepTimeMs < 1 || stepTimeMs > 255) throw new Error("stepTimeMs cabe em 1 byte");

  const tipo = COMPRESSAO[compressao];
  if (tipo === undefined) throw new Error(`compressão desconhecida: ${compressao}`);
  if (tipo === COMPRESSAO.zstd) throw new Error("zstd ainda não implementado — ver ADR 0008");

  // 1. Corta em blocos e comprime. Feito antes do cabeçalho porque só
  //    depois de comprimir se sabe o tamanho do índice.
  const blocos = [];
  if (tipo === COMPRESSAO.nenhuma) {
    blocos.push({ quadro: 0, bytes: dados });
  } else {
    const porBloco = quadrosPorBloco(canais, quadros);
    for (let q = 0; q < quadros; q += porBloco) {
      const fim = Math.min(quadros, q + porBloco);
      const cru = dados.subarray(q * canais, fim * canais);
      blocos.push({ quadro: q, bytes: await zlibComprime(cru) });
    }
  }

  // Sem compressão o índice não existe: o dado é um bloco contínuo.
  const nBlocos = tipo === COMPRESSAO.nenhuma ? 0 : blocos.length;

  // 2. Cabeçalhos variáveis.
  const vars = [];
  const texto = s => {
    const b = new TextEncoder().encode(s);
    const dado = new Uint8Array(b.length + 1);      // terminado em NUL
    dado.set(b);
    return dado;
  };
  if (midia) vars.push({ codigo: "mf", dado: texto(midia) });
  /* "bl" é código nosso. O FPP e o xLights ignoram código que não
     conhecem — só logam e seguem —, então isto não quebra compatibilidade. */
  if (rig) vars.push({ codigo: CODIGO_RIG, dado: texto(`${VERSAO_RIG};${rig.ch};${rig.fp}`) });
  const varBytes = vars.reduce((s, v) => s + VAR_CABECALHO + v.dado.length, 0);

  const tamCabecalho = CABECALHO + nBlocos * BLOCO_INDICE + 0 * FAIXA_ESPARSA;
  const offsetDados = arredonda4(tamCabecalho + varBytes);
  if (offsetDados > 0xFFFF) throw new Error("cabeçalho não cabe em 16 bits");

  const corpo = blocos.reduce((s, b) => s + b.bytes.length, 0);
  const saida = new Uint8Array(offsetDados + corpo);
  const dv = new DataView(saida.buffer);

  // 3. Cabeçalho fixo.
  saida.set([0x50, 0x53, 0x45, 0x51], 0);          // "PSEQ"
  dv.setUint16(4, offsetDados, true);
  saida[6] = 0;                                     // versão menor
  saida[7] = 2;                                     // versão maior
  dv.setUint16(8, tamCabecalho, true);
  dv.setUint32(10, canais, true);
  dv.setUint32(14, quadros, true);
  saida[18] = stepTimeMs;
  saida[19] = 0;
  saida[20] = ((nBlocos >> 4) & 0xF0) | tipo;
  saida[21] = nBlocos & 0xFF;
  saida[22] = 0;                                    // sem faixas esparsas
  saida[23] = 0;
  dv.setBigUint64(24, uniqueId ?? BigInt(Date.now()) * 1000n, true);

  // 4. Índice de blocos.
  let p = CABECALHO;
  for (let i = 0; i < nBlocos; i++) {
    dv.setUint32(p, blocos[i].quadro, true);
    dv.setUint32(p + 4, blocos[i].bytes.length, true);
    p += BLOCO_INDICE;
  }

  // 5. Cabeçalhos variáveis. O tamanho declarado inclui os 4 bytes
  //    de tamanho e código — é a convenção do escritor do FPP.
  for (const v of vars) {
    dv.setUint16(p, VAR_CABECALHO + v.dado.length, true);
    saida[p + 2] = v.codigo.charCodeAt(0);
    saida[p + 3] = v.codigo.charCodeAt(1);
    saida.set(v.dado, p + 4);
    p += VAR_CABECALHO + v.dado.length;
  }

  // 6. Dado de canal.
  p = offsetDados;
  for (const b of blocos) { saida.set(b.bytes, p); p += b.bytes.length; }
  return saida;
}

/**
 * Lê um .fseq V2. Serve de teste do escritor e de porta de entrada
 * pra arquivo vindo do xLights.
 * @returns {Promise<{canais:number,quadros:number,stepTimeMs:number,
 *   compressao:string,uniqueId:bigint,midia:string|null,dados:Uint8Array}>}
 */
export async function lerFseq(bytes) {
  const u8 = bytes instanceof Uint8Array ? bytes : new Uint8Array(bytes);
  if (u8.length < CABECALHO) throw new Error("arquivo curto demais pra ser fseq");
  const dv = new DataView(u8.buffer, u8.byteOffset, u8.byteLength);
  if (String.fromCharCode(u8[0], u8[1], u8[2], u8[3]) !== "PSEQ")
    throw new Error("não é um arquivo fseq");
  if (u8[7] !== 2) throw new Error(`versão ${u8[7]} não suportada, só a V2`);

  const offsetDados = dv.getUint16(4, true);
  const canais = dv.getUint32(10, true);
  const quadros = dv.getUint32(14, true);
  const stepTimeMs = u8[18];
  const tipo = u8[20] & 0x0F;
  const nBlocos = (((u8[20] & 0xF0) << 4) | u8[21]) >>> 0;
  const nEsparsas = u8[22];
  const uniqueId = dv.getBigUint64(24, true);
  if (tipo === COMPRESSAO.zstd) throw new Error("zstd ainda não implementado — ver ADR 0008");

  let p = CABECALHO;
  const indice = [];
  for (let i = 0; i < nBlocos; i++) {
    const quadro = dv.getUint32(p, true);
    const tam = dv.getUint32(p + 4, true);
    if (tam > 0) indice.push({ quadro, tam });
    p += BLOCO_INDICE;
  }
  p += nEsparsas * FAIXA_ESPARSA;

  let midia = null, rig = null;
  while (p + VAR_CABECALHO <= offsetDados) {
    const tam = dv.getUint16(p, true);
    if (tam < VAR_CABECALHO || p + tam > offsetDados) break;
    const codigo = String.fromCharCode(u8[p + 2], u8[p + 3]);
    const txt = new TextDecoder().decode(u8.subarray(p + 4, p + tam)).replace(/\0+$/, "");
    if (codigo === "mf") midia = txt;
    if (codigo === CODIGO_RIG) {
      const [v, ch, fp] = txt.split(";");
      // Carimbo de versão futura: melhor ignorar do que fingir que entendeu
      // e reprovar um arquivo bom.
      if (Number(v) === VERSAO_RIG && fp) rig = { ch: Number(ch), fp };
    }
    p += tam;
  }

  let dados;
  if (tipo === COMPRESSAO.nenhuma) {
    dados = u8.subarray(offsetDados, offsetDados + canais * quadros);
  } else {
    const partes = [];
    let off = offsetDados;
    for (const b of indice) {
      partes.push(await zlibDescomprime(u8.subarray(off, off + b.tam)));
      off += b.tam;
    }
    dados = new Uint8Array(partes.reduce((s, x) => s + x.length, 0));
    let q = 0;
    for (const parte of partes) { dados.set(parte, q); q += parte.length; }
  }
  if (dados.length !== canais * quadros)
    throw new Error(`dado tem ${dados.length} bytes, cabeçalho promete ${canais * quadros}`);

  return { canais, quadros, stepTimeMs, compressao: NOME_COMPRESSAO[tipo],
           uniqueId, midia, rig, dados };
}

/**
 * O arquivo foi renderizado pra este croqui? Avisar é o ponto: tocar um
 * fseq velho não dá erro nenhum, só manda pan pro canal de gobo.
 * @param {{rig:object|null,canais:number}} seq  saída de lerFseq
 * @param {{ch:number,fp:string}} atual          saída de impressaoDoRig
 */
export function conferirRig(seq, atual) {
  if (!seq.rig)
    return { ok: true, aviso: "arquivo sem carimbo de croqui — não dá pra conferir" };
  if (seq.rig.ch !== atual.ch)
    return { ok: false, aviso:
      `arquivo tem ${seq.rig.ch} canais, o croqui de agora tem ${atual.ch}` };
  if (seq.rig.fp !== atual.fp)
    return { ok: false, aviso:
      "mesmo número de canais, mas o croqui mudou — reexporte antes de usar" };
  return { ok: true, aviso: null };
}

/** Um quadro específico, já fatiado. Conveniência pra teste e inspeção. */
export function quadroDe(seq, i) {
  return seq.dados.subarray(i * seq.canais, (i + 1) * seq.canais);
}
